autocomplete/builders/common.lua

```diff
@@ -393,54 +408,102 @@ end
 -- Package compilation
 --
 
+---@alias packageType
+---|"class"
+---|"function"
+---|"method"
+---|"value"
+---|"lib"
+---|"event"
+
 --- @class exampleTable
 --- @field title string|nil The example title.
 --- @field description string|nil The description of the example.
 
 --- @class package
+--- @field name string
+--- @field useDefault boolean
+--- @field readOnly boolean
+--- @field optional boolean? Is this parameter optional?
+--- @field default string|integer|nil
+--- @field experimental boolean
+--- @field description string?
 --- @field key string The name of the file that generated this package.
---- @field type string The type definition for the package.
+--- @field type packageType The type definition for the package.
 --- @field folder string The folder that the package was created from.
 --- @field parent package The package this package is a child of.
 --- @field namespace string The full namespace of the package.
 --- @field deprecated boolean Allows marking definitions as deprecated. Those definitions aren't written to the web documentation.
 --- @field examples table<string, exampleTable>|nil A table containing the examples. Keys are the example's name/path to the example file.
 
+---@class packageValue : package
+---@field valuetype string
+---@field link string
+
+---@class packageFunctionArgument : package
+---@field tableParams packageFunctionArgument[]
```

Comment:
How does this recursive definition work?

Reply:
This is needed for the typing to be consistent with the `writeArgument` function in `mkdocs.lua`. Specifically, the code block
```lua
if (argument.tableParams) then
	for _, tableArg in ipairs(argument.tableParams) do
		writeArgument(file, tableArg, "\t" .. indent)
	end
end
```
says that everything in `argument.tableParams` must have the same type as `argument`. Although in practice, probably no argument will have a nested `tableParams` structure. 

I should change the type of `tableParams` to be `packageFunctionArgument[]|nil`. I'll do this in the next commit.

Reply:
Still, recursive annotation doesn't make this quite clear. What do you think if you had:
```lua
-- An entry in tableParams of a function
---@class packageFunctionArgumentSimple : package

---@class packageFunctionArgumentComplex : package
---@field tableParams packageFunctionArgumentSimple[]

-- Parameter or return value for a `packageFunction` or `packageMethod`.
---@alias packageFunctionArgument packageFunctionArgumentSimple|packageFunctionArgumentComplex
```

Reply:
It might be because I've gone too deep into Haskell, but I personally think the recursive definition is a bit easier to read than the one that uses `packageFunctionArgumentSimple` and `packageFunctionArgumentComplex`. But I can make that change if you think it's best.

Reply:
Well, I guess I look at things differently 😀. Would you still use the `packageFunctionArgumentSimple|packageFunctionArgumentComplex` union (feel free to change the name of the annotations if you think of something better), since recursive annotations are non-existent in our ecosystem?

Reply:
I think I've just had my brain broken by a little too much set theory. But it's probably best to avoid the recursive definitions when possible. 

I made the change, but used `packageFunctionArgument.Complex` and `packageFunctionArgument.Simple` (i.e. with `.` separators), because I thought it might be a helpful way to indicate these classes are very closely tied to the `packageFunctionArgument` class. But I can remove the dots if you think it's better without them.

Reply:
The dots are very good addition.